File: src/carmine/compiler/structures/Environment.java
package carmine.compiler.structures;

import java.util.HashMap;

public class Environment {
    private Environment enclosing = null;
    private HashMap<String, Object> variables = new HashMap<>();

    public void put(String name, Object value)
    {
        variables.put(name, value);
    }

    public boolean contains(Token token)
    {
        return variables.containsKey(token.getLexeme());
    }

    public Object get(String name)
    {
        if (variables.containsKey(name))
        {
            return variables.get(name);
        }

        throw new RuntimeException("Unknown variable: " + name);
    }

    public Object get(Token token)
    {
        if (variables.containsKey(token.getLexeme()))
        {
            return variables.get(token.getLexeme());
        }

        throw new RuntimeException(token.getLine() + " Unknown variable: " + token.getLexeme());
    }

    public void addEnclosing(Environment enclosing)
    {
        this.enclosing = enclosing;
    }

    public Environment getEnclosing()
    {
        return enclosing;
    }

    public HashMap<String, Object> getVariables()
    {
        return variables;
    }
}

File: src/carmine/compiler/structures/Environment.kt
package carmine.compiler.structures;

import java.util.HashMap;
class Environment {
    private var enclosing : Environment? = null
    private val  variables = HashMap<String, Any>()

    fun put(name : String, value : Any)
    {
        variables[name] = value
    }

    fun contains(token : Token) : Boolean
    {
        return variables.containsKey(token.lexeme)
    }

    fun get(name : String) : Any?
    {
        if (variables.containsKey(name))
        {
            return variables[name]
        }

        throw RuntimeException("Unknown variable: $name")
    }

    fun get(token : Token) : Any?
    {
        if (variables.containsKey(token.lexeme))
        {
            return variables[token.lexeme]
        }

        throw RuntimeException("${token.line} Unknown variable: ${token.lexeme}")
    }

    fun addEnclosing(enclosing : Environment)
    {
        this.enclosing = enclosing
    }

    fun getEnclosing() : Environment?
    {
        return enclosing
    }

    fun getVariables() : HashMap<String, Any>
    {
        return variables
    }
}
